src/ProjectIdProvider.php

```diff
@@ -0,0 +1,206 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Auth;
+
+use DomainException;
+use Google\Auth\HttpHandler\HttpHandlerFactory;
+use GuzzleHttp\Psr7\Request;
+
+/**
+ * ProjectIdProvider encapsulates the behavior for determining a google-cloud project id.
+ *
+ * This class allows you to dynamically determine the application's project id. For example:
+ *
+ *   $provider = new Google\Auth\ProjectIdProvider();
+ *   $projectId = $provider->getProjectId();
+ *
+ */
+class ProjectIdProvider
+{
+    /**
+     * @var callable
+     */
+    private $httpHandler;
+    const METADATA_SERVICE = 'http://169.254.169.254/computeMetadata/v1/project/project-id';
+
+    /**
+     * ProjectIdProvider constructor.
+     * @param callable $httpHandler An optional callback which delivers psr7 request
+     */
+    public function __construct(callable $httpHandler = null)
+    {
+        $this->httpHandler = $httpHandler;
+    }
+
+    /**
+     * Determines the project id for the current project. If one cannot be determined, a DomainException
+     * is thrown. The project is determined from the following locations respectively:
+     *
+     *   1. A Credentials file supplied via the GOOGLE_APPLICATION_CREDENTIALS env variable
+     *   2. The default application credentials
+     *   3. The Google Cloud SDK configuration command
+     *   4. The App Engine Standard env variable
+     *   5. The Google Cloud Compute Engine Meta Data service
+     *
+     * @return string The current project's id.
+     * @throws DomainException
+     */
+    public function getProjectId()
+    {
+        // List of sources for determining a project id
+        $sources = [
+            [ $this, 'fromKeyFilePath'],
+            [ $this, 'fromApplicationDefaultCredentials' ],
+            [ $this, 'fromSdk' ],
+            [ $this, 'fromAppEngineStandard' ],
+            [ $this, 'fromComputeEngineMetaData' ],
+        ];
+
+        // Look through all of the sources until we find a project id
+        foreach ($sources as $callable) {
+            $id = $callable();
+            if ($id !== null) {
+                return $id;
+            }
+        }
+
+        // If a project id could not be found, throw an exception
+        throw new DomainException('Could not determine project id');
+    }
+
+    /**
+     * Check for a project id in the Key File specified in the GOOGLE_APPLICATION_CREDENTIALS
+     * environment variable
+     *
+     * @return string|null
+     */
+    private function fromKeyFilePath()
```

Comment:
Both this method and the next are part of the [ApplicationDefaultCredentials strategy](https://cloud.google.com/docs/authentication/production#providing_credentials_to_your_application). In my mind, this name implies a user-supplied keyFilePath.

Perhaps `fromADCEnv` and `fromADCWellKnownFile`, or combine the two into `fromApplicationDefaultCredentials` and use the same strategy as demonstrated in `ApplicationDefaultCredentials::getCredentials()`?

```php
$jsonKey = CredentialsLoader::fromEnv()
    ?: CredentialsLoader::fromWellKnownFile();
```